pkg/kubectl/apply/element.go

```diff
@@ -0,0 +1,337 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apply
+
+import (
+	"fmt"
+)
+
+// FieldMeta defines the strategy used to apply a Patch for an element
+type FieldMeta interface {
+	// GetFieldMergeType specifies how a field should be merged.  One of
+	// replace, merge, retainkeys
+	GetFieldMergeType() string
+
+	// GetFieldMergeKey specifies which fields to use in order to merge items
+	// in lists
+	GetFieldMergeKey() []string
+
+	// GetFieldType returns the openapi field type - e.g. primitive, array, map, type, reference
+	GetFieldType() string
+}
+
+// FieldMetaImpl implements FieldMeta
+type FieldMetaImpl struct {
+	// The type of merge strategy to use for this field
+	// maybe "merge", "replace" or "retainkeys"
+	MergeType string
```

Comment:
We have talked about this. In case you forget, point it out one more time:
"retainkeys" may appear with "merge", e.g.: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L2539